server/channels/store/sqlstore/post_store.go

```diff
@@ -924,11 +924,22 @@ func (s *SqlPostStore) InvalidateLastPostTimeCache(channelId string) {
 }
 
 //nolint:unparam
-func (s *SqlPostStore) GetEtag(channelId string, allowFromCache, collapsedThreads bool) string {
-	q := s.getQueryBuilder().Select("Id", "UpdateAt").From("Posts").Where(sq.Eq{"ChannelId": channelId}).OrderBy("UpdateAt DESC").Limit(1)
+func (s *SqlPostStore) GetEtag(channelId string, userId string, allowFromCache, collapsedThreads bool) string {
+	var q sq.SelectBuilder
 	if collapsedThreads {
-		q.Where(sq.Eq{"RootId": ""})
+		q = s.getQueryBuilder().
+			Select("Posts.Id", "GREATEST(Posts.UpdateAt, COALESCE(Threads.LastReplyAt, 0), COALESCE(ThreadMemberships.LastUpdated, 0)) AS UpdateAt").
+			From("Posts").
+			LeftJoin("Threads ON Threads.PostId = Posts.Id").
+			LeftJoin("ThreadMemberships ON ThreadMemberships.PostId = Posts.Id AND ThreadMemberships.UserId = ?", userId).
+			Where(sq.Eq{"Posts.ChannelId": channelId}).
+			Where(sq.Eq{"Posts.RootId": ""}).
+			OrderBy("GREATEST(Posts.UpdateAt, COALESCE(Threads.LastReplyAt, 0), COALESCE(ThreadMemberships.LastUpdated, 0)) DESC").
```

Comment:
I'm a bit concerned about the new JOINs and the GREATEST in the OrderBy. I'll need to check the query plan of both the old and new versions to see whether this is efficient enough.